src/beamlime/applications/handlers.py

```diff
@@ -0,0 +1,114 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+import pathlib
+from dataclasses import dataclass
+from typing import Any, NewType
+
+import plopp as pp
+import scipp as sc
+
+from beamlime.logging import BeamlimeLogger
+
+from ._workflow import Events, FirstPulseTime, Histogrammed, WorkflowPipeline
+from .base import BeamlimeMessage, HandlerInterface
+
+
+@dataclass
+class HistogramUpdated(BeamlimeMessage):
+    content: pp.graphics.basefig.BaseFig
+
+
+@dataclass
+class RawDataSent(BeamlimeMessage):
+    content: Events
+
+
+class DataReductionHandler(HandlerInterface):
+    """Data reduction handler to process the raw data and update the histogram.
+
+    It receives a list of events, and reduces them into a histogram.
+    It also triggers the update of a plot stream node.
+    """
+
+    pipeline: WorkflowPipeline
+
+    def __init__(self) -> None:
+        self.output_da: Histogrammed
+        self.stream_node: pp.Node
+        super().__init__()
+
+    def format_received(self, data: Any) -> str:
+        return f"{len(data)} pieces of {Events.__name__}"
+
+    def process_first_input(self, da: Events) -> None:
+        first_pulse_time = da[0].coords['event_time_zero'][0]
+        self.pipeline[FirstPulseTime] = first_pulse_time
+
+    def process_first_output(self, data: Histogrammed) -> None:
+        self.output_da = Histogrammed(sc.zeros_like(data))
+        self.info("First data as a seed of histogram: %s", self.output_da)
+        self.stream_node = pp.Node(self.output_da)
+        self.figure = pp.figure1d(
+            self.stream_node,
+            title="Wavelength Histogram",
+            grid=True,
+        )
+
+    async def process_data(self, data: Events) -> Histogrammed:
+        self.info("Received, %s", self.format_received(data))
+        self.pipeline[Events] = data
+        return self.pipeline.compute(Histogrammed)
+
+    async def process_message(
+        self, message: RawDataSent | BeamlimeMessage
```

Comment:
I think the annotation is equivalent to `message: BeamlimeMessage` because `RawDataSent` is a subclass of `BeamlimeMessage`.

Reply:
I wanted to annotate it as ``RawDataSet`` so that it's easy to see that this method is meant to handle ``RawDataSent`` and then pylint was not happy about it...
I'll just keep it as ``BeamlimeMessage`` for now.